consensus/ibft/signer/signer.go

```diff
@@ -0,0 +1,336 @@
+package signer
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/0xPolygon/polygon-edge/crypto"
+	"github.com/0xPolygon/polygon-edge/types"
+	"github.com/0xPolygon/polygon-edge/validators"
+)
+
+var (
+	ErrEmptyCommittedSeals        = errors.New("empty committed seals")
+	ErrEmptyParentCommittedSeals  = errors.New("empty parent committed seals")
+	ErrInvalidCommittedSealLength = errors.New("invalid committed seal length")
+	ErrInvalidCommittedSealType   = errors.New("invalid committed seal type")
+	ErrRepeatedCommittedSeal      = errors.New("repeated seal in committed seals")
+	ErrNonValidatorCommittedSeal  = errors.New("found committed seal signed by non validator")
+	ErrNotEnoughCommittedSeals    = errors.New("not enough seals to seal block")
+	ErrSignerMismatch             = errors.New("mismatch address between signer and message sender")
+	ErrValidatorNotFound          = errors.New("not found validator in validator set")
+	ErrInvalidValidatorSet        = errors.New("invalid validator set type")
+	ErrInvalidSignature           = errors.New("invalid signature")
+	ErrNilParentHeader            = errors.New("parent header is nil")
+)
+
+// Signer is responsible for signing for blocks and messages in IBFT
+type Signer interface {
+	Type() validators.ValidatorType
+	Address() types.Address
+	InitIBFTExtra(header *types.Header, parentCommittedSeal Sealer, set validators.Validators) error
+	GetIBFTExtra(header *types.Header) (*IstanbulExtra, error)
+	GetParentCommittedSeals(header *types.Header) (Sealer, error)
+	WriteProposerSeal(*types.Header) (*types.Header, error)
+	SignIBFTMessage([]byte) ([]byte, error)
+	Ecrecover([]byte, []byte) (types.Address, error)
+	EcrecoverFromHeader(*types.Header) (types.Address, error)
+	CreateCommittedSeal([]byte) ([]byte, error)
+	VerifyCommittedSeal(validators.Validators, types.Address, []byte, []byte) error
+	WriteCommittedSeals(*types.Header, map[types.Address][]byte) (*types.Header, error)
+	VerifyCommittedSeals(
+		header *types.Header,
+		validators validators.Validators,
+		quorumSize int,
+	) error
+	VerifyParentCommittedSeals(
+		parent, header *types.Header,
+		parentValidators validators.Validators,
+		quorum int,
+		mustExist bool,
+	) error
+	CalculateHeaderHash(*types.Header) (types.Hash, error)
+}
+
+type SignerImpl struct {
+	keyManager KeyManager
+}
+
+func NewSigner(
+	keyManager KeyManager,
+) Signer {
+	return &SignerImpl{
+		keyManager: keyManager,
+	}
+}
+
+func (s *SignerImpl) Type() validators.ValidatorType {
+	return s.keyManager.Type()
+}
+
+func (s *SignerImpl) Address() types.Address {
+	return s.keyManager.Address()
+}
+
+func (s *SignerImpl) InitIBFTExtra(
+	header *types.Header,
+	parentCommittedSeal Sealer,
+	validators validators.Validators,
+) error {
+	putIbftExtra(header, &IstanbulExtra{
+		Validators:          validators,
+		ProposerSeal:        nil,
+		CommittedSeal:       s.keyManager.NewEmptyCommittedSeal(),
+		ParentCommittedSeal: parentCommittedSeal,
+	})
+
+	return nil
+}
```

Comment:
Method returns an error in the signature, but no implementation (this is the only one) returns an error

Reply:
Fixed at a44e68ccd0022172a59f4d9cc9e7f1adbd957fbf